c7n/resources/ebs.py

```diff
@@ -218,6 +218,77 @@ def process(self, snapshots, event=None):
         return resources
 
 
+@Snapshot.action_registry.register('copy-volume-tags')
+class SnapshotCopyVolumeTags(BaseAction):
+    """
+    Copy EBS Volume tags to snapshots
+
+    By default, skip-missing-volumes is True. To break
+    policy execution if a volume is not found, set
+    skip-missing-volumes to False
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: ebs-snapshot-copy-vol-tags
+            resource: ebs-snapshot
+            actions:
+              - type: copy-volume-taags
+                skip-missing-volumes: True
+                tags:
+                    - Application
+                    - Owner
+    """
+
+    permissions = ('ec2:DescribeVolumes', 'ec2:CreateTags')
+    schema = type_schema(
+        'copy-volume-tags',
+        required=['tags'],
+        **{
+            'skip-missing-volumes': {'type': 'boolean'},
+            'tags': {'type': 'array', 'items': {'type': 'string'}}
+        }
+    )
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('ec2')
+        tag_keys = self.data.get('tags', [])
+        volume_ids = [r['VolumeId'] for r in resources]
+
+        volume_tag_map = self.get_volume_tag_map(volume_ids)
+
+        for r in resources:
+            if r['VolumeId'] not in volume_tag_map:
+                self.log.warning(
+                    'Unable to find volume: %s for snapshot: %s' % (r['VolumeId'], r['SnapshotId']))
+                continue
+            vol_tags = volume_tag_map[r['VolumeId']]
+            if vol_tags:
+                add_tags = [{'Key': t['Key'], 'Value': t['Value']} for t in vol_tags if t['Key'] in tag_keys] # noqa
+            else:
+                continue
+            if add_tags:
+                vol_tags.extend(add_tags)
```

Comment:
The volume tags should be intersected with the snapshot tags to find a delta of tag values to change, else as is, everytime this runs its going to set the same tag values on a snapshot that already has those values.